File: appcore/src/main/java/mobi/cangol/mobile/http1/BinaryHttpResponseHandler.java


package mobi.cangol.mobile.http1;

import android.content.Context;
import android.os.Message;
import android.util.Log;

import java.io.IOException;
import java.io.UnsupportedEncodingException;

import okhttp3.Response;
import okhttp3.ResponseBody;

public class BinaryHttpResponseHandler extends AsyncHttpResponseHandler {
    // Allow images by default
    private String[] mAllowedContentTypes = new String[]{
            "application/zip",
            "application/x-zip-compressed",
            "application/octet-stream"
    };

    /**
     * Creates a new BinaryHttpResponseHandler
     */
    public BinaryHttpResponseHandler() {
        super();
    }

    /**
     * Creates a new BinaryHttpResponseHandler with context
     */
    public BinaryHttpResponseHandler(Context context) {
        super(context);
    }

    /**
     * Creates a new BinaryHttpResponseHandler, and overrides the default allowed
     * content types with passed String array (hopefully) of content types.
     */
    public BinaryHttpResponseHandler(String[] allowedContentTypes) {
        this();
        mAllowedContentTypes = allowedContentTypes;
    }

    /**
     * Creates a new BinaryHttpResponseHandler, and overrides the default allowed
     * content types with passed String array (hopefully) of content types.
     */
    public BinaryHttpResponseHandler(Context context, String[] allowedContentTypes) {
        this(context);
        mAllowedContentTypes = allowedContentTypes;
    }


    //
    // Callbacks to be overridden, typically anonymously
    //

    /**
     * Fired when a request returns successfully, override to handle in your own code
     *
     * @param binaryData the body of the HTTP response from the server
     */
    public void onSuccess(byte[] binaryData) {
    }

    /**
     * Fired when a request returns successfully, override to handle in your own code
     *
     * @param statusCode the status code of the response
     * @param binaryData the body of the HTTP response from the server
     */
    public void onSuccess(int statusCode, byte[] binaryData) {
        onSuccess(binaryData);
    }

    /**
     * Fired when a request fails to complete, override to handle in your own code
     *
     * @param error      the underlying cause of the failure
     * @param binaryData the response body, if any
     * @deprecated
     */
    public void onFailure(Throwable error, byte[] binaryData) {
        // By default, call the deprecated onFailure(Throwable) for compatibility
        onFailure(error);
    }


    //
    // Pre-processing of messages (executes in background threadpool thread)
    //

    protected void sendSuccessMessage(int statusCode, byte[] responseBody) {
        sendMessage(obtainMessage(SUCCESS_MESSAGE, new Object[]{statusCode, responseBody}));
    }

    //
    // Pre-processing of messages (in original calling thread, typically the UI thread)
    //

    protected void handleSuccessMessage(int statusCode, byte[] responseBody) {
        onSuccess(statusCode, responseBody);
    }

    protected void handleFailureMessage(Throwable e, byte[] responseBody) {
        onFailure(e, responseBody);
    }

    protected void handleFailureMessage(Throwable e, String responseBody) {
        byte[] response = null;
        if (responseBody != null) {
            try {
                onFailure(e, responseBody.getBytes("utf-8"));
            } catch (UnsupportedEncodingException e1) {
                Log.d("UnsupportedEncoding", e.getMessage());
            }
        } else {
            onFailure(e, response);
        }
    }

    // Methods which emulate android's Handler and Message methods
    protected void handleMessage(Message msg) {
        Object[] response;
        switch (msg.what) {
            case SUCCESS_MESSAGE:
                response = (Object[]) msg.obj;
                handleSuccessMessage(((Integer) response[0]).intValue(), (byte[]) response[1]);
                break;
            case FAILURE_MESSAGE:
                response = (Object[]) msg.obj;
                if (response[1] instanceof byte[]) {
                    handleFailureMessage((Throwable) response[0], (byte[]) response[1]);
                } else {
                    handleFailureMessage((Throwable) response[0], (String) response[1]);
                }
                break;
            default:
                super.handleMessage(msg);
                break;
        }
    }
    void sendResponseMessage(Response response) {
        Log.d(">>", "sendResponseMessage=");
        ResponseBody responseBody=response.body();
        if(response.isSuccessful()){
            String contentType=response.body().contentType().toString();
            boolean foundAllowedContentType = false;
            for (String anAllowedContentType : mAllowedContentTypes) {
                if (anAllowedContentType.equals(contentType)) {
                    foundAllowedContentType = true;
                }
            }
            if (!foundAllowedContentType) {
                sendFailureMessage(new IOException("Content-Type not allowed! "+contentType), response.message());
                return;
            }
            try {
                sendSuccessMessage(response.code(), responseBody.bytes());
            } catch (IOException e) {
                sendFailureMessage(e, (byte[]) null);
            }
        }else{
            sendFailureMessage(new IOException("code="+response.code()), response.message());
        }
    }
}

File: appcore/src/main/java/mobi/cangol/mobile/http/BinaryHttpResponseHandler.kt
package mobi.cangol.mobile.http

import android.content.Context
import android.os.Message
import android.util.Log
import okhttp3.Response
import java.io.IOException
import java.io.UnsupportedEncodingException

open class BinaryHttpResponseHandler : AsyncHttpResponseHandler {
    // Allow images by default
    private var mAllowedContentTypes = arrayOf("application/zip", "application/x-zip-compressed", "application/octet-stream")

    /**
     * Creates a new BinaryHttpResponseHandler
     */
    constructor() : super() {}

    /**
     * Creates a new BinaryHttpResponseHandler with context
     */
    constructor(context: Context) : super(context) {}

    /**
     * Creates a new BinaryHttpResponseHandler, and overrides the default allowed
     * content types with passed String array (hopefully) of content types.
     */
    constructor(allowedContentTypes: Array<String>) : this() {
        mAllowedContentTypes = allowedContentTypes
    }

    /**
     * Creates a new BinaryHttpResponseHandler, and overrides the default allowed
     * content types with passed String array (hopefully) of content types.
     */
    constructor(context: Context, allowedContentTypes: Array<String>) : this(context) {
        mAllowedContentTypes = allowedContentTypes
    }


    //
    // Callbacks to be overridden, typically anonymously
    //

    /**
     * Fired when a request returns successfully, override to handle in your own code
     *
     * @param binaryData the body of the HTTP response from the server
     */
    fun onSuccess(binaryData: ByteArray) {
        //do nothings
    }

    /**
     * Fired when a request returns successfully, override to handle in your own code
     *
     * @param statusCode the status code of the response
     * @param binaryData the body of the HTTP response from the server
     */
    fun onSuccess(statusCode: Int, binaryData: ByteArray) {
        onSuccess(binaryData)
    }

    /**
     * Fired when a request fails to complete, override to handle in your own code
     *
     * @param error      the underlying cause of the failure
     * @param binaryData the response body, if any
     */
    fun onFailure(error: Throwable, binaryData: ByteArray) {
        // By default, call the deprecated onFailure(Throwable) for compatibility
        onFailure(error)
    }


    //
    // Pre-processing of messages (executes in background threadpool thread)
    //

    protected fun sendSuccessMessage(statusCode: Int, responseBody: ByteArray) {
        sendMessage(obtainMessage(SUCCESS_MESSAGE, arrayOf<Any>(statusCode, responseBody)))
    }

    //
    // Pre-processing of messages (in original calling thread, typically the UI thread)
    //

    protected fun handleSuccessMessage(statusCode: Int, responseBody: ByteArray) {
        onSuccess(statusCode, responseBody)
    }

    protected fun handleFailureMessage(e: Throwable, responseBody: ByteArray) {
        onFailure(e, responseBody)
    }

    override fun handleFailureMessage(e: Throwable, responseBody: String) {
        if (responseBody != null) {
            try {
                onFailure(e, responseBody.toByteArray(charset("utf-8")))
            } catch (e1: UnsupportedEncodingException) {
                Log.d("UnsupportedEncoding", e.message)
            }

        } else {
            onFailure(e, null as ByteArray)
        }
    }

    // Methods which emulate android's Handler and Message methods
    override fun handleMessage(msg: Message) {
        val response: Array<Any>
        when (msg.what) {
            SUCCESS_MESSAGE -> {
                response = msg.obj as Array<Any>
                handleSuccessMessage((response[0] as Int).toInt(), response[1] as ByteArray)
            }
            FAILURE_MESSAGE -> {
                response = msg.obj as Array<Any>
                if (response[1] is ByteArray) {
                    handleFailureMessage(response[0] as Throwable, response[1] as ByteArray)
                } else {
                    handleFailureMessage(response[0] as Throwable, response[1] as String)
                }
            }
            else -> super.handleMessage(msg)
        }
    }

    internal override fun sendResponseMessage(response: Response) {
        Log.d(">>", "sendResponseMessage=")
        val responseBody = response.body()
        if (response.isSuccessful) {
            val contentType = response.body()!!.contentType()!!.toString()
            var foundAllowedContentType = false
            for (anAllowedContentType in mAllowedContentTypes) {
                if (anAllowedContentType == contentType) {
                    foundAllowedContentType = true
                }
            }
            if (!foundAllowedContentType) {
                sendFailureMessage(IOException("Content-Type not allowed! $contentType"), response.message())
                return
            }
            try {
                sendSuccessMessage(response.code(), responseBody!!.bytes())
            } catch (e: IOException) {
                sendFailureMessage(e, null)
            }

        } else {
            sendFailureMessage(IOException("code=" + response.code()), response.message())
        }
    }
}